STM/vertx/README.adoc

```diff
@@ -0,0 +1,179 @@
+
+What it's good for:
+
+  - horizontal scaling by using better hardware so that more threads can do be used to service the
+    workload;
+
+Usage:
+------
+
+  mvn exec:java -Pdemo2
+
+or
+
+  java -cp target/stm-vertx-demo-5.6.0.Final-SNAPSHOT-fat.jar demo.demo12.PersistentSharedTheatreVerticle
+
+Start a second verticle that will share the same STM object. In the output of the above command
+look for the line that reports the uid of the STM object, for example:
+
+> Theatre STM uid: 0:ffffac1182c6:9197:5912d4ff:1
+
+The actual uid may change from run to run. Pass this uid the second JVM using a java system property:
+
+  mvn exec:java -Pdemo2 -Duid=0:ffffac1182c6:9197:5912d4ff:1
+
+or
+
+  java -cp target/stm-vertx-demo-5.6.0.Final-SNAPSHOT-fat.jar demo.demo12.PersistentSharedTheatreVerticle 0:ffffac1182c6:9197:5912d4ff:1
+
+Create two bookings using services running in the two different JVMs (on endpoints 8080 and 8082
+respectively):
+
+  curl -X POST http://localhost:8080/api/theatre/Apollo
+  curl -X POST http://localhost:8082/api/theatre/Savoy
+
+Check that each JVM reports the correct number of bookings (namely 2):
+
+  curl -X GET http://localhost:8080/api/theatre
+  curl -X GET http://localhost:8082/api/theatre
+
+Use Case 3: Managing shared state across different STM objects
+===============================================
+
+Characteristics:
+  - RECOVERABLE and EXCLUSIVE
+  - trip service verticle (multiple instances) using STM objects to maintain theatre and taxi bookings.
+  - providing HTTP endpoints for making trip, theatre and taxi bookings.
+
+What it's good for:
+  - composing transactional operations across different STM objects 
+
+The trip service fulfils booking requests by updating shared STM objects representing the theatre and
+taxi booking services respectively.
+
+  mvn exec:java -Pdemo3
+
+or
+
+  java -cp target/stm-vertx-demo-5.6.0.Final-SNAPSHOT-fat.jar demo.demo3.TripSTMVerticle
+
+Make two trip bookings:
+
+  curl -X POST http://localhost:8080/api/trip/Savoy/ABC
+  curl -X POST http://localhost:8080/api/trip/Apollo/XYZ
+
+and a single theatre booking:
+
+  curl -X POST http://localhost:8080/api/theatre/Savoy
+
+observe that each booking is serviced by a different verticle. Check that number of theatre (3) and
+number of taxi (2) bookings are correct:
+
+  curl -X GET http://localhost:8080/api/theatre
+  curl -X GET http://localhost:8080/api/taxi
+
+Use Case 4: Stress Testing:
+--------------------------
+
+Non of the previous use cases demonstrate contention of the STM objects. This next case will start
+use case 1 (theatre booking example) and then make lots of concurrent trip bookings and validate that
+the expected number of bookngs are made.
+
+Usage:
+------
+
+Start the theatre service in one window or in the background:
+
+  mvn exec:java -Pdemo1
+
+and now make lots of concurrent bookings:
+
+  mvn exec:java -Pstress
```

Comment:
This is windows.

It's strange how yours doesn't build the code (whereas my output does show it doing something like that)

$ ./build.sh -f STM/vertx/pom.xml exec:java -Pstress                                                                                                                                                                                                                 
Maven already exists                                                                                                                                                                                                                                                 
Maven is correct version                                                                                                                                                                                                                                             
./tools/maven/bin/mvn -s tools/maven/conf/settings.xml install  '-f' 'STM/vertx/pom.xml' 'exec:java' '-Pstress'                                                                                                                                                      
[INFO] Scanning for projects...                                                                                                                                                                                                                                      
[INFO]                                                                                                                                                                                                                                                               
[INFO] ------------------------------------------------------------------------                                                                                                                                                                                      
[INFO] Building Narayana - STM and vertx demo 5.6.0.Final-SNAPSHOT                                                                                                                                                                                                   
[INFO] ------------------------------------------------------------------------                                                                                                                                                                                      
[INFO]                                                                                                                                                                                                                                                               
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ stm-vertx-demo ---                                                                                                                                                                             
[WARNING] Using platform encoding (Cp1252 actually) to copy filtered resources, i.e. build is platform dependent!                                                                                                                                                    
[INFO] Copying 1 resource                                                                                                                                                                                                                                            
[INFO]                                                                                                                                                                                                                                                               
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ stm-vertx-demo ---                                                                                                                                                                                  
[INFO] Nothing to compile - all classes are up to date                                                                                                                                                                                                               
[INFO]                                                                                                                                                                                                                                                               
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ stm-vertx-demo ---                                                                                                                                                                     
[WARNING] Using platform encoding (Cp1252 actually) to copy filtered resources, i.e. build is platform dependent!                                                                                                                                                    
[INFO] skip non existing resourceDirectory C:\Users\still\Documents\jbosstm\quickstart\STM\vertx\src\test\resources                                                                                                                                                  
[INFO]                                                                                                                                                                                                                                                               
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ stm-vertx-demo ---                                                                                                                                                                          
[INFO] No sources to compile                                                                                                                                                                                                                                         
[INFO]                                                                                                                                                                                                                                                               
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ stm-vertx-demo ---                                                                                                                                                                                     
[INFO] No tests to run.                                                                                                                                                                                                                                              
[INFO]                                                                                                                                                                                                                                                               
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ stm-vertx-demo ---                                                                                                                                                                                               
[INFO]                                                                                                                                                                                                                                                               
[INFO] --- maven-shade-plugin:2.3:shade (default) @ stm-vertx-demo ---                                                                                                                                                                                               
[WARNING] Map in class org.apache.maven.plugins.shade.resource.ManifestResourceTransformer declares value type as: class java.util.jar.Attributes but saw: class java.lang.String at runtime                                                                         
[INFO] Including io.vertx:vertx-core:jar:3.3.3 in the shaded jar.                                                                                                                                                                                                    
[INFO] Including io.netty:netty-common:jar:4.1.5.Final in the shaded jar.                                                                                                                                                                                            
[INFO] Including io.netty:netty-buffer:jar:4.1.5.Final in the shaded jar.                                                                                                                                                                                            
[INFO] Including io.netty:netty-transport:jar:4.1.5.Final in the shaded jar.                                                                                                                                                                                         
[INFO] Including io.netty:netty-handler:jar:4.1.5.Final in the shaded jar.                                                                                                                                                                                           
[INFO] Including io.netty:netty-codec:jar:4.1.5.Final in the shaded jar.                                                                                                                                                                                             
[INFO] Including io.netty:netty-handler-proxy:jar:4.1.5.Final in the shaded jar.                                                                                                                                                                                     
[INFO] Including io.netty:netty-codec-socks:jar:4.1.5.Final in the shaded jar.                                                                                                                                                                                       
[INFO] Including io.netty:netty-codec-http:jar:4.1.5.Final in the shaded jar.                                                                                                                                                                                        
[INFO] Including io.netty:netty-codec-http2:jar:4.1.5.Final in the shaded jar.                                                                                                                                                                                       
[INFO] Including io.netty:netty-resolver:jar:4.1.5.Final in the shaded jar.                                                                                                                                                                                          
[INFO] Including io.netty:netty-resolver-dns:jar:4.1.5.Final in the shaded jar.                                                                                                                                                                                      
[INFO] Including io.netty:netty-codec-dns:jar:4.1.5.Final in the shaded jar.                                                                                                                                                                                         
[INFO] Including com.fasterxml.jackson.core:jackson-core:jar:2.7.4 in the shaded jar.                                                                                                                                                                                
[INFO] Including com.fasterxml.jackson.core:jackson-databind:jar:2.7.4 in the shaded jar.                                                                                                                                                                            
[INFO] Including com.fasterxml.jackson.core:jackson-annotations:jar:2.7.0 in the shaded jar.                                                                                                                                                                         
[INFO] Including io.vertx:vertx-web:jar:3.3.3 in the shaded jar.                                                                                                                                                                                                     
[INFO] Including io.vertx:vertx-auth-common:jar:3.3.3 in the shaded jar.                                                                                                                                                                                             
[INFO] Including org.jboss.narayana.stm:stm:jar:5.6.0.Final-SNAPSHOT in the shaded jar.                                                                                                                                                                              
[INFO] Including org.jboss.narayana:common:jar:5.6.0.Final-SNAPSHOT in the shaded jar.                                                                                                                                                                               
[INFO] Including org.jboss.narayana.arjunacore:arjuna:jar:5.6.0.Final-SNAPSHOT in the shaded jar.                                                                                                                                                                    
[INFO] Including org.jboss.narayana.arjunacore:txoj:jar:5.6.0.Final-SNAPSHOT in the shaded jar.                                                                                                                                                                      
[INFO] Including org.jboss.logging:jboss-logging:jar:3.1.4.GA in the shaded jar.                                                                                                                                                                                     
[INFO]                                                                                                                                                                                                                                                               
[INFO] --- maven-install-plugin:2.4:install (default-install) @ stm-vertx-demo ---                                                                                                                                                                                   
[INFO] Installing C:\Users\still\Documents\jbosstm\quickstart\STM\vertx\target\stm-vertx-demo-5.6.0.Final-SNAPSHOT.jar to C:\Users\still\.m2\repository\org\jboss\narayana\quickstart\stm\stm-vertx-demo\5.6.0.Final-SNAPSHOT\stm-vertx-demo-5.6.0.Final-SNAPSHOT.jar
[INFO] Installing C:\Users\still\Documents\jbosstm\quickstart\STM\vertx\pom.xml to C:\Users\still\.m2\repository\org\jboss\narayana\quickstart\stm\stm-vertx-demo\5.6.0.Final-SNAPSHOT\stm-vertx-demo-5.6.0.Final-SNAPSHOT.pom                                       
[INFO]                                                                                                                                                                                                                                                               
[INFO] --- exec-maven-plugin:1.4.0:java (default-cli) @ stm-vertx-demo ---                                                                                                                                                                                           
Waiting for 1000 requests                                                                                                                                                                                                                                            
0 out of 1000 requests failed in 7014 ms                                                                                                                                                                                                                             
[WARNING] thread Thread[Thread-3,5,demo.util.Stress] was interrupted but is still alive after waiting at least 15000msecs                                                                                                                                            
[WARNING] thread Thread[Thread-3,5,demo.util.Stress] will linger despite being asked to die via interruption                                                                                                                                                         
[WARNING] NOTE: 1 thread(s) did not finish despite being asked to  via interruption. This is not a problem with exec:java, it is a problem with the running code. Although not serious, it should be remedied.                                                       
[WARNING] Couldn't destroy threadgroup org.codehaus.mojo.exec.ExecJavaMojo$IsolatedThreadGroup[name=demo.util.Stress,maxpri=10]                                                                                                                                      
java.lang.IllegalThreadStateException                                                                                                                                                                                                                                
        at java.lang.ThreadGroup.destroy(ThreadGroup.java:778)                                                                                                                                                                                                       
        at org.codehaus.mojo.exec.ExecJavaMojo.execute(ExecJavaMojo.java:328)                                                                                                                                                                                        
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)                                                                                                                                                         
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)                                                                                                                                                                           
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)                                                                                                                                                                           
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)                                                                                                                                                                           
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)                                                                                                                                                  
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)                                                                                                                                                   
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)                                                                                                                                     
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)                                                                                                                                                                   
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:355)                                                                                                                                                                                            
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:155)                                                                                                                                                                                              
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)                                                                                                                                                                                                  
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:216)                                                                                                                                                                                                   
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:160)                                                                                                                                                                                                     
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)                                                                                                                                                                                               
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)                                                                                                                                                                             
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)                                                                                                                                                                     
        at java.lang.reflect.Method.invoke(Method.java:498)                                                                                                                                                                                                          
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)                                                                                                                                                                       
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)                                                                                                                                                                               
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)                                                                                                                                                                     
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)                                                                                                                                                                                 
[INFO] ------------------------------------------------------------------------                                                                                                                                                                                      
[INFO] BUILD SUCCESS                                                                                                                                                                                                                                                 
[INFO] ------------------------------------------------------------------------                                                                                                                                                                                      
[INFO] Total time: 18.430 s                                                                                                                                                                                                                                          
[INFO] Finished at: 2017-05-11T11:38:31+01:00                                                                                                                                                                                                                        
[INFO] Final Memory: 24M/384M                                                                                                                                                                                                                                        
[INFO] ------------------------------------------------------------------------